Configuration file for pkg-config and hide ts module

mshabunin opened the pull request:
This PR contains commit from #3882 (which is port of #3305) and additional changes:
- pkg-config template now contains `Libs.private` field with static dependencies
- macro `ocv_get_all_libs` determine libraries kind more carefully
- added sample makefile with pkg-config call (can be used on buildbot)
- cmake output: added dependencies information
- cmake output: fixed GDAL library line in case when it is not found

Also:
- marked `ts` module as `INTERNAL` -> it will not be installed

mshabunin commented:
@tSed, @SpecLad, @alalek, could you, please, take a look?

SpecLad commented on cmake/OpenCVModule.cmake:
```diff
@@ -780,11 +780,13 @@ macro(_ocv_create_module)
-    LIBRARY DESTINATION ${OPENCV_LIB_INSTALL_PATH} COMPONENT libs
-    ARCHIVE DESTINATION ${OPENCV_LIB_INSTALL_PATH} COMPONENT dev
-    )
+  # if(";${OPENCV_MODULES_PUBLIC};" MATCHES ";${the_module};")
```
Is this supposed to be commented out?

SpecLad commented on samples/cpp/example_cmake/Makefile:
```diff
@@ -0,0 +1,12 @@
+CXX ?= g++
+
+CXXFLAGS += -c -Wall `pkg-config --cflags opencv`
+LDFLAGS += `pkg-config --libs --static opencv`
```
I think it would be better to use `$(shell ...)` here, to only query this stuff once per Makefile. It's not very important for _this_ Makefile, but it would set a better example.

SpecLad commented on cmake/OpenCVGenPkgconfig.cmake:
```diff
@@ -8,71 +8,89 @@
+        string(REGEX REPLACE "^lib" "" libname ${libname})
+      endif()
+      set(libprefix "-l")
+    elseif(item MATCHES "^-[WlL]" OR item MATCHES "^\\$")
```
What's `-W`? Aren't those warning options?

mshabunin replied:
It was left accidentally, I'll remove it. Same for `OR item MATCHES "^\\$"`

SpecLad commented on cmake/OpenCVGenPkgconfig.cmake:
```diff
@@ -8,71 +8,89 @@
+        string(REGEX REPLACE "^liblib" "lib" libname ${libname})
+      else()
+        string(REGEX REPLACE "^lib" "" libname ${libname})
+      endif()
```
Don't both branches do the same thing?

mshabunin replied:
cmake removes all occurencies: for pattern "^lib" it will transform "liblibwebp" to "webp", but actual library is named "liblibwebp" (when compiled from OpenCV source tree).

SpecLad replied:
Huh. Sounds like a bug in CMake, but okay.

SpecLad commented:
It would be nice if you could port the pkg-config changes from 6e121b2e2, too. (Adding support for imported targets and removing support for `-L`).

mshabunin commented:
All problems have been fixed, pkg-config-related changes from 6e121b2 have been ported too.

mshabunin commented:
I've fixed one issue with imported targets handling: _ippicv_ library is used as imported target, thus it's original location was also added to linker flags, like this: "-L<path-to-opencv-sources>/<path-to-extracted-ippicv>".

alalek commented on cmake/OpenCVUtils.cmake:
```diff
@@ -796,11 +796,20 @@ macro(ocv_get_all_libs _modules _extra _3rdparty)
-          list(INSERT ${_3rdparty} 0 ${dep})
+          get_target_property(_isimported ${dep} IMPORTED)
+          # hack for IPPICV
+          if (NOT _isimported OR ${dep} STREQUAL "ippicv")
```
`IPPICV` is static library. Is it included into shared build configs?

mshabunin replied:
No, in GenPkgConfig file, there is a block:
```.cpp
if (BUILD_SHARED_LIBS)
  set(OPENCV_PC_LIBS_PRIVATE "${_extra}")
else()
  set(OPENCV_PC_LIBS_PRIVATE
    "-L\${exec_prefix}/${OPENCV_3P_LIB_INSTALL_PATH}"
    "${_3rdparty}"
    "${_extra}"
  )
 endif()
```
it means, that _3rdparty_ libraries (deployed with OpenCV) are not included to the list for shared build.

This concrete check was made to correctly separate _ippicv_ from other libraries, maybe it is better to change this line to `if(NOT _isimported OR "${item}" MATCHES "3rdparty")` as it was made in GenPkgConfig.

alalek commented on cmake/OpenCVUtils.cmake:
```diff
@@ -796,11 +796,20 @@ macro(ocv_get_all_libs _modules _extra _3rdparty)
       if (NOT DEFINED OPENCV_MODULE_${dep}_LOCATION)
         if (TARGET ${dep})
-          list(INSERT ${_3rdparty} 0 ${dep})
+          get_target_property(_isimported ${dep} IMPORTED)
```
It is very strange check. There are no correlation between exported targets (you need this) and imported targets. Exported 3rdparty targets are defined via line like this:
```
ocv_install_target(${ZLIB_LIBRARY} EXPORT OpenCVModules ARCHIVE DESTINATION ${OPENCV_3P_LIB_INSTALL_PATH} COMPONENT dev)
```

I'm not sure that we can read contents of `EXPORT OpenCVModules`, but we can update wrapper `ocv_install_target` to build this list.

mshabunin replied:
IPP library is added like this:
```.cmake
add_library(ipp${name} STATIC IMPORTED)
<...>
install(FILES ${ipp_lib} DESTINATION ${OPENCV_3P_LIB_INSTALL_PATH} COMPONENT main)
```
Other 3rdparty libraries are normal targets and they are automatically put into _3rdparty_ list because of `OPENCV_MODULE_${dep}_LOCATION` variable absence. But IPP library is imported target and it will be put into _extra_ list without ths check.

It seems that we can use `OpenCVModules_TARGETS` variable to check whether the target is exported, but there will be no IPP library in this list (I guess).

tSed commented on cmake/OpenCVGenPkgconfig.cmake:
```diff
@@ -8,71 +8,82 @@
+      list(APPEND _lst "-L${libdir}" "-l${libname}")
+    else()
+      list(APPEND _lst "-l${item}")
+    endif()
```
Could item be "-L/some/libdir" or "-Wl,..."?
If yes, how are these case handled?

mshabunin replied:
I've partially ported commit  6e121b2 from branch 2.4. That commit removed the `-L` options appearance for CUDA libraries (other part will be ported soon, I believe).
The '-Wl' option is not possible, I used it for experiments with different linking flags on early stages without success.

SpecLad replied:
BTW, adding linker options to the library list is non-portable, and breaks the build in various ways. I specifically removed `-L` support to discourage adding them.

mshabunin commented:
@tSed, it makes same thing with 4 additional lines, why do you think it is better?
Also, I've already changed these lines: instead of _"is not imported"_ there is _"is exported"_ check now. As @alalek wrote, this expression makes better sense, because all 3rdparty targets are exported with `EXPORT OpenCVModules` option and are stored in variable `OpenCVModules_TARGETS`. `ippicv` is still an exception from this mechanism, but I don't think it is a serious issue.

```.cmake
        if (TARGET ${dep})
          # 3rdparty libraries: targets exported to OpenCVModules + ippicv
          list(FIND OpenCVModules_TARGETS "${dep}" _isexported)
          if (_isexported GREATER "-1" OR "${dep}" STREQUAL "ippicv")
            list(INSERT ${_3rdparty} 0 ${dep})
          else()
            list(INSERT ${_extra} 0 ${dep})
          endif()
        else()
          list(INSERT ${_extra} 0 ${dep})
        endif()
```

tSed commented:
@mshabunin, 4 explicit additional lines, it just make the  specific handling clearer and easier to remove if ipp gets more "standard" some day... just a maintainance point of view.
Yeah, the _"is not imported"_ / _"is exported"_ thing, just saw it while finishing the review. I'm just too familiar with squashed and rebased patch series. 

(Note that having patch adding stuff, stuff that is removed in the following patch does not make the review process easy and is quite misleading :-/)

mshabunin commented:
I've rebased and squashed all commits.

3rdparty target check has been changed again, now we compare __ARCHIVE_OUTPUT_DIRECTORY__ target property with __3P_LIBRARY_OUTPUT_PATH__. _ippicv_ specific handling has been extracted to separate block.

tSed commented:
lgtm :+1

Tested in cross-compilation with buildroot tree [1] and this defconfig [2].
Without the last commit [3] which adds the 2 patches of this PR), vlc failes to link because of wrong absolute path in the _"Libs"_ field of the ```opencv.pc```; after vlc successfully builds and links.

[1] https://github.com/tSed/buildroot/tree/sma/opencv-3 
[2] https://gist.github.com/tSed/bf0ad2727658685f9ff3
[3] https://github.com/tSed/buildroot/commit/1ca112459de82ce62e331fa2b980e78e92cc4953

alalek commented:
:+1:

paroj commented:
what was the reason of hiding ts again? it is quite useful for testing modules depending on opencv.

mshabunin commented:
@paroj, it was fix for the issue: http://code.opencv.org/issues/3225

paroj commented:
@mshabunin, commented on the issue:
a different fix could be:
change
```
# define GTEST_API_ CV_EXPORTS
```
to
```
# define GTEST_API_
```
i.e. not exporting the gtest symbols in the first place. this would still allow exporting the ts module without exporting the GTEST API.
